must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallTransfer.java

```diff
@@ -0,0 +1,46 @@
+package org.checkerframework.checker.mustcall;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.type.TypeMirror;
+import org.checkerframework.dataflow.analysis.RegularTransferResult;
+import org.checkerframework.dataflow.analysis.TransferInput;
+import org.checkerframework.dataflow.analysis.TransferResult;
+import org.checkerframework.dataflow.cfg.node.StringConcatenateAssignmentNode;
+import org.checkerframework.dataflow.cfg.node.StringConcatenateNode;
+import org.checkerframework.framework.flow.CFAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+
+/**
+ * Transfer function for the must-call type system. Handles defaulting for string concatenations.
+ */
+public class MustCallTransfer extends CFTransfer {
+
+  private MustCallAnnotatedTypeFactory atypeFactory;
+
+  public MustCallTransfer(CFAnalysis analysis) {
+    super(analysis);
+    atypeFactory = (MustCallAnnotatedTypeFactory) analysis.getTypeFactory();
+  }
+
+  @Override
+  public TransferResult<CFValue, CFStore> visitStringConcatenate(
+      StringConcatenateNode n, TransferInput<CFValue, CFStore> input) {
+    return handleStringConcatenation(super.visitStringConcatenate(n, input));
+  }
+
+  @Override
+  public TransferResult<CFValue, CFStore> visitStringConcatenateAssignment(
+      StringConcatenateAssignmentNode n, TransferInput<CFValue, CFStore> in) {
+    return handleStringConcatenation(super.visitStringConcatenateAssignment(n, in));
+  }
+
+  private TransferResult<CFValue, CFStore> handleStringConcatenation(
```

Comment:
Please document to aid in understanding.  This sets all string concatenations to `@CalledMethods({})`.
Why is that necessary in the transfer function?  That is, why cannot the AnnotatedTypeFactory do this?

Reply:
> That is, why cannot the AnnotatedTypeFactory do this?

It could. I borrowed this implementation from the Value Checker, which also has custom handling for string concatenation (e.g. [here](https://github.com/typetools/checker-framework/blob/master/framework/src/main/java/org/checkerframework/common/value/ValueTransfer.java#L590)).

I think doing this in the transfer is more performant than doing it as part of the ATF, but the effect is the same - the type of all String concatenations is bottom. The reason to do so is that implicitly calling `toString()` on an object with a must-call obligation doesn't create a corresponding must-call obligation for the string. Without this code, the below would not pass the OCC's must-call check because the argument to `println` would need to call `close` (which it can't, because it's a string!):

```
void test(@Owning Socket s) {
    System.out.println("this is my socket: " + s);
    s.close();
}
```

Obviously that code does call `close`, so the warning is a false positive.